### PR TITLE
Introduce dismiss event for text-expander-element

### PR DESCRIPTION
**Pull request by iulia-b**
This PR is adding `dismiss` method on the Public API of TextExpanderElement. Calling this method will dismiss the existing menu provided through the `provider` and setting the lookbackIndex to the cursor position to avoid triggering it when the user continues typing. 

This will fix a few of the items specified in https://github.com/github/issues/issues/1119, mostly the ones related to existing suggestions flow. 
- `Escape` behavior is handled here since it is useful to have once with the introduction of multi-word
- dismissal on pressing ArrowLeft/Right, as well as onClick anywhere outside the menu feels more specific to Github behaviour so it will be implemented on dotcom side (it will use this event)
- a `##` sequence will trigger dismissal, also on dotcom side since it is a MD specific behaviour

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +90,13 @@ expander.addEventListener('text-expander-value', function(event) {
 })
 ```
 
+**`text-expander-dismiss`** is an event the text-expander-element is listening to. The event can be fired in order to hide the menu provided through `provide` method. It is not expecting any parameter. It is used internally to hide the menu when Escape key it hit.
```
Comment (lukewar) [inline README.md:93]: > It is used internally to hide the menu when Escape key it hit.

I am confused by this sentence. AFAIK `text-expander-element` already dismisses menu on `Esc` key so this event is not needed for that. My understanding was that this event was introduced to dismiss the menu by some external factors. Should we get rid of that sentence?

```suggestion
**`text-expander-dismiss`** is an event the text-expander-element is listening to. The event can be fired in order to hide the menu provided through `provide` method. It is not expecting any parameter.
```

File: README.md [hunk 2]
```diff
@@ -90,6 +90,13 @@ expander.addEventListener('text-expander-value', function(event) {
...
+
+```js
+const expander = document.querySelector('text-expander')
+expander.dispatchEvent('text-expander-dismiss')
+
```
Comment (keithamus) [inline README.md:98]: If you have the element in reference, you can call `dismiss()`, as it is part of the public API of `text-expander`.

I'm not sure we need to use events for this mechanism.

File: README.md (hunk 1 above)
Comment (iulia-b) [reply README.md:93]: Do you mean expose through this text or in the code? 
If you mean through the text, & you are referring to `It is used internally to hide the menu when Escape key it hit` : not sure it is needed, i did it mainly because it is stopping propagation of Esc and this could be useful information for other users of this lib

File: README.md (hunk 2 above)
Comment (lvivski) [reply README.md:98]: I think I'm a bit confused here. `text-expander` element seems to only have `.keys` as public interface https://github.com/github/text-expander-element/blob/760da4af27410c93a9cf3e7ad782d77a01987a34/src/text-expander-element.ts#L210-L233 . I agree though, that having something like `.dismiss()` can be helpful. Do you suggest it instead of the event?

Comment (iulia-b) [reply README.md:98]: I checked with @keithamus and confirmed this ^ I had the same confusion. Added dismiss() and updated the test (tks Keith for the suggestion)

File: README.md (hunk 1 above)
Comment (iulia-b) [reply README.md]: resolving this as i changed the implementation and removed the dismiss event

**Comment by chanakyabhardwajj** [comment]
The PR description needs to be updated to remove `text-expander-dismiss` event.

File: src/text-expander-element.ts [hunk 3]
```diff
@@ -199,7 +203,7 @@ class TextExpander {
 
   onKeydown(event: KeyboardEvent) {
     if (event.key === 'Escape' && (this.menu || this.combobox)) {
```
Comment (chanakyabhardwajj) [inline src/text-expander-element.ts:205]: Suggestion - Instead of checking for the presence of `menu` or `combobox` you can delegate this to the `deactivate` method and return a boolean from it. For reference see this commit - https://github.com/github/text-expander-element/pull/19/commits/cd81b9a36b6f2c951c29ab7bff4e2e71ded18c7b

Comment (iulia-b) [reply src/text-expander-element.ts]: cherry-picked it back 👍

File: src/text-expander-element.ts [hunk 4]
```diff
@@ -86,16 +86,21 @@ class TextExpander {
     this.combobox.navigate(1)
   }
 
-  deactivate() {
+  deactivate(cursor?: number) {
```
Comment (lukewar) [inline src/text-expander-element.ts:89]: How tightly connected is the cursor with deactivation logic? I have concerns if it should be part of this function. For example in `onCommit()` we are setting `this.lookBackIndex` again on line 133. In `onKeydown()` we are passing both `this.input.selectionEnd || this.lookBackIndex` while `this.lookBackIndex` is used in the body of that function. Did you consider having separate function for dealing with cursor updates?

Comment (iulia-b) [reply src/text-expander-element.ts:89]: I did consider adding it separate, i think there was a reason i had to pass it here. But good point, i'll try to de-couple it and set the lookbackIndex separately in each function

Comment (lukewar) [reply src/text-expander-element.ts:89]: Yeah, if there was some edge case I would not insist on the refactor, I suggest though to add a test checking it or at least a comment that will give some context about this edge case.

Comment (iulia-b) [reply src/text-expander-element.ts]: I changed it and could find any edge case. I hope i am not missing something

Comment (iulia-b) [reply src/text-expander-element.ts]: Ah, i think the problem will arise when dismiss will be triggered from outside of the webComponent (like https://github.com/github/github/pull/163808) . Yes that was the issue. it needs to be set to the current cursor position when it is triggered from outside the component

Comment (lukewar) [reply src/text-expander-element.ts]: Good point. It could be controlled though inside `dismiss()` implementation I guess.

File: test/text-expander-element-test.js [hunk 5]
```diff
@@ -47,6 +47,27 @@ describe('text-expander element', function() {
       const {key} = event.detail
       assert.equal(':', key)
     })
+
+    it('Escape dismisses the menu', async function() {
```
Comment (koddsson) [inline test/text-expander-element-test.js:51]:
```suggestion
    it('dismisses the menu when dismiss() is called', async function() {
```

File: src/query.ts [hunk 6]
```diff
@@ -21,10 +21,9 @@ export default function query(
   const keyIndex = text.lastIndexOf(key, cursor - 1)
   if (keyIndex === -1) return
 
-  if (multiWord) {
-    // Stop matching at the lookBackIndex
-    if (keyIndex < lookBackIndex) return
+  if (keyIndex < lookBackIndex) return
```
Comment (koddsson) [inline src/query.ts:24]: Keep the comment?

```suggestion
  // Stop matching at the lookBackIndex	
  if (keyIndex < lookBackIndex) return
```

**Review by koddsson — APPROVED** [review]
This looks good to me 👍🏻

**Review by keithamus — APPROVED** [review]
Pair reviewed with @koddsson on this. I'm happy with this too.
